V2.1.0

SimonGLM opened the pull request:
# Changes

### DAEMON:
* Updates:
  - restructured data logging system: 
   - introduced log rotation and possibility to enable/disable data logging
   - simplified structure of data storage path under /var/muondetector
   - auto delete old data and log files
  - added geo position managing system with three modes of operation:
    - static: report position that was statically set (by user or as a result of a lock-in operation, see below)
    - auto: report position from most recent uBlox position fix
    - lock-in: fix a uBlox position only when accuracy is better than a user defined limit (for pos accuracy and pos DOP values) and make this the new static position
  - added several position filters applicable to auto and lock-in modes:
    - no filter: uBlox position fixes are not filtered (bad accuracy and bad precision))
    - kalman: apply a kalman filter which smoothes the series of last fixes efficiently but poorly filters long term transients/drifts (poor accuracy, excellent precision)
    - histogram mean (good accuracy, fair precision)
    - histogram median (better accuracy, fair precision)
    - histogram most probable value (mpv) (best accuracy, fair precision)
 
* Bugfixes:
  - removed wrong QT connect call which resulted in a warning in log messages after daemon startup
  - fixed reconnect to MQTT broker after connection was lost temporarily
  - fixed bug in I2C subsystem which lead to a stuck bus after certain read operations and preventing detection of a device in the next logical read access


### GUI
* Updates
  - Extended the geographic map feature with position marker (zoom level independent), semitransparent accuracy indicator and info box with current coordinates and position accuracy.
  - Added daemon hardware and software info under "Parameters" tab.
  - Added time as scan parameter in "Scans" tab.
  - Added display of log rotation interval and event log enable flag in "Log" tab.
  - Fixed bug which prevented auto update of histograms and plots.
  - Optimized color schemes to give better appearance for dark themes
  -  changed timing input selection from radio button group to drop down menu
  - Parameters: Added daemon hardware and software information
  - Scans: Added time as scan parameter
  - Log: Added display of log rotation interval and event log enable flag
  - Map: Intensively improved the geographic map
    - position marker (zoom level independent)
    - semitransparent accuracy indicator
    - info box with current coordinates and position accuracy
    - controls of geo position mode (auto, static, lock-in), static position and lock-in limits

* Bugfixes:
  - Fixed bug which prevented auto update of histograms and plots
  - Rate buffer was sometimes sent with incorrect size when connecting with the GUI leading to occasional crashes

hangeza committed:
implemented transmission and display of extended MQTT connection state

hangeza committed:
removed boolean mqtt status entirely from daemon

hangeza committed:
changed color of MQTT "Connecting" status field in GUI from blue to yellow

hangeza committed:
small mod in code for MQTT status panel in GUI

hangeza committed:
removed blank lines

hangeza committed:
removed blank lines

hangeza committed:
fixed bug which prevented proper parsing of UBX messages

hangeza committed:
fixed bug where the endianness of message rate config messages was wrong

hangeza committed:
Merge branch 'merge-refactor-ublox' into refactor-ublox

hangeza committed:
few mods to make it compile after the last merge

hangeza committed:
improved method for setting ublox dynamic model

hangeza committed:
replaced hardcoded UBX message IDs by named constants

hangeza committed:
failsafe fixed for unknown UBX message IDs in GUI

hangeza committed:
Intensively improved the geographic map feature of the GUI.
- red position marker with constant size now visible in all zoom levels of the map
- semitransparent blue circle indicating the position accuracy (radius depending on zoom level)
- additional info box in upper left corner displaying the current coordinates and pos accuracy
- map widget set inactive when no detector connected

hangeza committed:
removed remnant debug output

hangeza committed:
- completed migration of QtSerialUblox class wrt refactoring using the templated get<T> function
- migrated potentially dangerous pointer to last acknowledged message to std::unique_ptr

hangeza committed:
updated changelog for GUI

hangeza committed:
made the satellite table in GUI sortable

hangeza committed:
applied clang-format

hangeza committed:
GNSS sat widget in GUI: satellites are additionaly marked with a secondary circle when used in fix

hangeza committed:
separated all ublox-related structs and constants from general muondetector structs

hangeza committed:
Merge branch 'refactor-ublox' into dev

hangeza committed:
Merge pull request #71 from MuonPi/dev

Dev

hangeza committed:
Merge pull request #72 from MuonPi/refactor-ublox

Merge pull request #71 from MuonPi/dev

hangeza committed:
removed blank line

hangeza committed:
applied clang-format

hangeza committed:
overhaul of GnnsPosWidget class: modernized language and removed potential out-of-bounds access of vector GNSS_COLORS

hangeza committed:
reformatted widgets in parameter monitor tab in GUI

hangeza committed:
added some more tcp message codes for future utilization

hangeza committed:
applied clang-format

hangeza committed:
renamed GpsSatsForm into GnssInfoForm and modernized its language

hangeza committed:
applied clang-format

hangeza committed:
renamed Settings widget into more descriptive UbloxSettingsForm

hangeza committed:
applied clang-format

hangeza committed:
removed malformed whitespaces

hangeza committed:
removed uneccesary code fragments and blank line

hangeza committed:
removed getpass() function from main and renamed it in muondetector-login since the function's name gave rise to a bogus security warning according to hi-cpp

hangeza committed:
reordered all streaming operators into custom_io_operators.h/cpp

hangeza committed:
applied clang-format

hangeza committed:
removed using namespace std in qtserialublox.h

hangeza committed:
extended struct UbxMessage and moved handling of UbxMessages away from Qt to STL

hangeza committed:
updated Histogram class to conform with modern C++

hangeza committed:
small bugfix in handling the nr of target interfaces in UBX_CFG_CFG message

hangeza committed:
made some local variables const

hangeza committed:
applied clang-format

hangeza committed:
forgot to add a new source file

hangeza committed:
adapted code to exterminate 3 compiler warnings

hangeza committed:
removed more compiler warnings (signedness warnings)

hangeza committed:
moved 2 static constants from global scope as members into QtSerialUblox class

hangeza committed:
commented out two currently unused variables

hangeza committed:
applied clang-format

hangeza committed:
fixed preprocessor switches for selecting code for the right Qt version

hangeza committed:
changed base for qt version debug output back to 16

hangeza committed:
modified the widget layouts and sizes slightly to allow for a slightly smaller window height

hangeza committed:
fixed broken layout in parameter monitoring tab

CodingBobby committed:
fixed text color

When switching the theme of the OS to dark, the text often becomes white to stay visible on the dark background. For areas where the background color is fixed, this automatic change should not happen. To fix this, the text colour (css tag "color") is now manually set to black in certain places.

hangeza committed:
added mqtt inhibit function on daemon side

hangeza committed:
Merge branch 'dev' of github.com:MuonPi/muondetector into dev

hangeza committed:
Merge pull request #75 from CodingBobby/dev

Fix text color for dark themes

hangeza committed:
changed colorschemes of following widgets to respect the system's color scheme depending on theme:
- MQTT connection panel
- timepulse panel
- sat position widget
- I/Q display widget (on GNSS tab)

hangeza committed:
applied clang-format

hangeza committed:
adapted now all color schemes to automatically adjust to the system's color theme

hangeza committed:
applied clang-format

hangeza committed:
small bugfix for last commit: replot of I/Q data widget on GNSS tab was not done automatically

hangeza committed:
bugfix: TCP request MSG_BIAS_SWITCH_REQUEST was not answered by daemon and MSG_BIAS_VOLTAGE_REQUEST was returning the bias switch status instead of the bias dac voltage

hangeza committed:
small mod in sending bias switch status

hangeza committed:
added mqtt inhibit function and checkbox in GUI

hangeza committed:
updated some code to be more readable

hangeza committed:
bugfix: mqtt status was not sent after mqtt inhibit was set/unset

hangeza committed:
migrated GNSS string constants to pure STL code

hangeza committed:
migrated more GNSS constants to std language

hangeza committed:
changed all Gnss related structs to pure std idiom

hangeza committed:
bugfix: latest changes in Gnss structs not compilable due to missing cast

hangeza committed:
bugfix: parsing of some fields of GNSS satellite UBX message was flawed for ublox8 devices

marvin5300 committed:
Added cmake configuration for mosquitto library in windows.

hangeza committed:
optimized parsing and transformation of pseudorange residual GNSS parameter

hangeza committed:
Merge branch 'dev' of github.com:MuonPi/muondetector into dev

hangeza committed:
updated GnssSatellite struct: renamed members and changed types of some members to reflect the source types from ublox messages better

marvin5300 committed:
Added some new container files and started optimizing signals and slots in daemon.

hangeza committed:
cleaned up Histogram class at several places

hangeza committed:
added include directory /usr/local/include for OSX builds

hangeza committed:
added required include

hangeza committed:
applied fix to two depracated warnings caused by Qt versions since 5.15

hangeza committed:
fixed wrong declaration syntax and defined empty method which resulted in linker error when missing.

hangeza committed:
bugfix: added type casts in streamer functions for GnssSatellite since the values were scrambled after the members of the class were changed to std types

hangeza committed:
forgot to add type cast for one field in last commit

hangeza committed:
added templated put function for placing integer values of arbitrary type into UBX messag

hangeza committed:
commented out one apparently unused line of code which generated a compiler warning

hangeza committed:
fixed warnings caused by allocation of static buffers with variable size

hangeza committed:
migrated two ublox structs to std types

hangeza committed:
small changes to improve readability and to suffice modern language standards

hangeza committed:
Merge pull request #76 from MuonPi/dev

Dev

hangeza committed:
reverted all changes from commit 9317a0aa0d5716ca9045123ebf22745613e8aeea (in branch cleanup-daemon)

hangeza committed:
removed one artefakt definition left from previous reversion

hangeza committed:
Merge pull request #77 from MuonPi/fix-cleanup-branch

Fix cleanup branch

hangeza committed:
removed all Qt references from gpio definitions on daemon side, such as GPIO_PINMAP

hangeza committed:
removed all Qt content from gpio signal and pin descriptors, GUI side

hangeza committed:
optimized iterator-based for-loop into for-range loop

hangeza committed:
optimized status.h/cpp in GUI:
- values for all parameters are reset at disconnect such that the values from previous connection are not preserved
- non-member variable rateSecondsBuffered moved as member into Status class and initialized with constexpr default value

hangeza committed:
applied clang-format

hangeza committed:
Consistently use TIMING_MUX_SELECTION type for timing input instead of selection index. Changed timing selection in GUI from QButtonGroup to drop-down menu

hangeza committed:
applied clang-format

hangeza committed:
Merge pull request #78 from MuonPi/gpio-system-to-std

Gpio system to std

hangeza committed:
updated changelogs

hangeza committed:
Merge branch 'position-models' of github.com:MuonPi/muondetector into dev

hangeza committed:
removed some variable doubling in daemon

hangeza committed:
bugfix: removed one remnant paranthese

hangeza committed:
restructured geo position handling and started implementation of geo position model and states

hangeza committed:
added handling of lock-in and static position states

hangeza committed:
- Implemented lock-in mode which observes histogram mean values of coordinates and fixes them once the given accuracy is reached.
- Rewrite of Property class as template.
- Deactivated code for Kalman filtering for now.

hangeza committed:
Merged definition of position mode and static position into struct PositionModeConfig. Implemented transmission of PositionModeConfig object via TCP socket on daemon side.

hangeza committed:
started implementation of position mode configuration on GUI side

hangeza committed:
added functionality to set pos config and static position in GUI

hangeza committed:
modified geographic map to show also altitude

hangeza committed:
added two more properties to PositionModeConfig struct, i.e. lock-in constraint parameters

hangeza committed:
applied clang-format

hangeza committed:
commented out one line of code (which shall be repatriated in the future) and promoted one qDebug output to qInfo

hangeza committed:
replaced position lock-in constraint constants by member variables of PositionModeConfig

hangeza committed:
added edit fields for pos config properties in GUI

hangeza committed:
initializing edit fields for pos config in GUI

hangeza committed:
Added struct FixType which contains enum for GNSS fix type and description strings.
Geo position is now transmitted with correct errors in case no fix could be achieved for longer time (was zero before).

hangeza committed:
removed unnecessary cyclic transmission of geo model struct via TCP socket

hangeza committed:
removed whitespaces in fix-type description strings

hangeza committed:
added sub struct FilterType to PositionModeConfig and renamed the "name" member to accomodate also strings describing the filter type

hangeza committed:
forgot to add one line for sending updated position config status after lock-in completed

hangeza committed:
bugfix attempt for seldom MQTT connection failures, where the connection turns to "disconnected" and no new connection is established

hangeza committed:
Merge pull request #80 from MuonPi/dev

Dev

hangeza committed:
Merge branch 'position-models' of github.com:MuonPi/muondetector into position-models

hangeza committed:
added reset method to kalman filter class

hangeza committed:
added field for filter configuration of position model

hangeza committed:
implemented setting of filter type in position config in map widget of the GUI

marvin5300 committed:
Added qml Signal for right click mouse event coordinates.

hangeza committed:
send static position in case the position model is changed since this is not done automatically

hangeza committed:
bugfix: lock-in mode with filters other than histo_mean was still unnecessarily waiting to lock until the histograms are filled with large statistics

hangeza committed:
Merge branch 'position-models' of github.com:MuonPi/muondetector into position-models

hangeza committed:
transmit static position on new connection and in log intervals

hangeza committed:
added pointer to libconfig file configuration to Configuration struct in daemon in order to potentially write config data back to the config file

hangeza committed:
implemented takeover of static position on right click event in map

hangeza committed:
increased numerical precision of geo coordinate strings in map

hangeza committed:
small bugfix: static pos was not set to valid when sending off from GUI

hangeza committed:
slightly modified one variable init and added debug output

hangeza committed:
Merge branch 'position-models' of github.com:MuonPi/muondetector into position-models

hangeza committed:
bugfix: transmission of static position was still not correct in static mode

hangeza committed:
added histogram for position DOP

hangeza committed:
position DOP now considered as criterion for lock-in

hangeza committed:
removed old code fragments

hangeza committed:
added histogram for time DOP

hangeza committed:
modified autoscale rules of histogram class a bit

hangeza committed:
added methods for calculation of median and most probable value in histogram class

hangeza committed:
added implementation of position filter methods histo_mean and histo_mpv

hangeza committed:
modified range on rescaling histogram

hangeza committed:
removed old code fragments

hangeza committed:
applied clang-format

hangeza committed:
removed unused constant

marvin5300 committed:
Fixed bug in MqttHandler where error_handler() was only called if new status differs from m_status.
This inhibits a desired reconnect loop.

marvin5300 committed:
Fixed stray } in mqtthandler.cpp

marvin5300 committed:
Removed unnecessary braces '{' '}' in gnssinfoform.cpp to remove a compiler warning.

marvin5300 committed:
Merge branch 'dev' into fix_mqtt_reconnect

marvin5300 committed:
Fixed implicit conversion from uint32_t to bool causing an error in qtserialublox_processmessages.cpp while compiling with clang-11.

marvin5300 committed:
Fixed MqttHandler. Removed old functions which did not work properly.
Configured mosquitto with an exponential reconnect backoff.
Let mosquitto run properly asynchronous.

marvin5300 committed:
Fixed login program.

hangeza committed:
Merge pull request #81 from MuonPi/fix_mqtt_reconnect

Fix mqtt reconnect

hangeza committed:
Merge pull request #82 from MuonPi/dev

Dev

marvin5300 committed:
Fixed muonpi login (please test).

hangeza committed:
changed default position mode to Auto

SimonGLM committed:
Removed whitespace in log/data-file headers.
Headers in file had two whitespace, while
the rows only had one ws as a separator.

SimonGLM committed:
Added GNSS::FixType::name_log_safe to fix space-separated csv-like reading errors

SimonGLM committed:
Encase swString in literal quotes to make string log-safe

SimonGLM committed:
Log messages now stored with one timestamp per log interval.
Fixes log intervals spreading over multiple seconds

SimonGLM committed:
Removed extra array
Changes according to review comments

SimonGLM committed:
Merge pull request #89 from MuonPi/fix-log-formatting

Fix log formatting -> improving machine readability

SimonGLM committed:
Initial commit on new release

SimonGLM committed:
Merge branch 'main' into dev

SimonGLM committed:
Merge branch 'dev' into v2.1.0

SimonGLM committed:
Changelog comments to be formulated

marvin5300 committed:
Started adding a network discovery class.

marvin5300 committed:
NetworkDiscovery is now functional:
Just create an object of this class both in daemon and gui and specify if it should be NetworkDiscovery::DeviceType::DAEMON or "GUI
and give it an udp port to use. Use the same port for both daemon and gui. Connect GUI button to "searchDevices()" function.
Connect "foundDevices(...)" signal to the list of available IP addresses (differentiate between types of devices given by quint16 in the QList of QPairs).
To be done:
 - Check if the primitive encoding of the device type is sufficient for suppressing rogue udp broadcasts of other programs to interfere with the functionality of this system.

marvin5300 committed:
Added the network discovery function to the gui and daemon.

marvin5300 committed:
Fixed a bug where NetworkDiscovery was not a pointer and was therefore deleted at the end of daemon constructor.

marvin5300 committed:
Added missing networkdiscovery.h to daemon.h

marvin5300 committed:
Removed debug line.

marvin5300 committed:
Merge branch 'dev' into v2.1.0

hangeza committed:
Merge branch 'dev' of github.com:MuonPi/muondetector into dev

hangeza commented on gui/src/mainwindow.cpp:
```diff
@@ -608,103 +612,101 @@ void MainWindow::receivedTcpMessage(TcpMessage tcpMessage)
+        emit daemonVersionReceived(hw_ver, sw_ver);
+        return;
+    } else {
+        qDebug() << "received unknown TCP message, msgID =" << QString::number(static_cast<int>(msgID));
```
this should rather be a warning, shouldn't it?
```suggestion
        qWarning() << "received unknown TCP message, msgID =" << QString::number(static_cast<int>(msgID));
```

hangeza reviewed (commented):
Some minor issues still to be changed and changelogs should be updated.

hangeza commented:
Be aware that the merge of the branch position-models into dev with PR #83 was reverted. Hence, the position-models functionality is NOT part of dev, v2.1.0 and specifically this PR.

hangeza committed:
Merge branch 'position-models' into dev

hangeza committed:
Merge branch 'dev' into position-models

SimonGLM commented:
@hangeza You're right! We should wait until that feature is ready to be merged.

hangeza committed:
added persistent storage of position mode and static pos to settings file

hangeza committed:
- use existing text descriptions of PositionModeConfig enums to write the coresponding entries into the settings files
- replaced one critical error message by a qFatal()
- made one const string static in Daemon::writeSettingsToFile()

hangeza committed:
removed blank line

hangeza committed:
deleted commented-out line

hangeza committed:
deleted commented-out code blocks

hangeza committed:
moved ublox uart timeout constant to common definitions in config.h

hangeza committed:
fixed wrong console output message

hangeza committed:
promoted debug output message to warning message

hangeza committed:
added files for GeoPosManager class

hangeza committed:
extended constructor of Histogram class by two additional arguments

hangeza committed:
added comparison operators to two structs

hangeza committed:
moved handling of GeoPosition from daemon into separate class GeoPosManager

hangeza committed:
removed unused math definition from daemon

hangeza committed:
applied clang-format

hangeza committed:
Merge branch 'dev' into position-models

hangeza committed:
splitted main function of GeoPosManager into smaller blocks to improve code quality

hangeza committed:
Merge pull request #91 from MuonPi/position-models

Position models functionality now fully implemented. Therefore merging to dev and closing the PR

hangeza committed:
Merge branch 'dev' into v2.1.0

SimonGLM committed:
Fixed "emit toConsole" not having newline in the end

hangeza committed:
removed obsolete declarations

hangeza committed:
added else to if clause to improve processing since only one signal is handled by this code at a time

hangeza committed:
Added RateBuffer class to potentially replace rate measurement and handling. An instance of RateBuffer is created for each GPIO input signal is connected to the PigpiodHandler::signal and thus tracking the GPIO rates.

hangeza committed:
cleaned up cmake file slightly and removed obsolete forward declaration

marvin5300 committed:
Fixed a bug where GUI would crash randomly on connecting to daemon due to gpio_rates not properly handled
in status.cpp.
Removed deprecated options "peerPort" and "peerAddress" and made sure networkdiscovery would use the udp port equivalent of our usual port.

hangeza committed:
Merge branch 'dev' of github.com:MuonPi/muondetector into dev

hangeza committed:
reverted some changes of last commit to be compatible with c++17 and older libc versions

hangeza committed:
bugfix: set serverPort instead of peerPort cmd line option

hangeza committed:
Merge branch 'dev' of github.com:MuonPi/muondetector into dev

hangeza committed:
applied clang-format

hangeza committed:
Fixed bug, where access to some I2C devices was not reliable as reported by some users. Now, device probing and identification should be more robust and should not get confused by single byte read failures, as well.

hangeza committed:
Merge branch 'dev' into v2.1.0

hangeza committed:
updated changelogs of daemon and GUI

hangeza committed:
optimized program flow of too complex function

hangeza committed:
applied clang-format

hangeza committed:
added rate buffer for ublox hardware counter

hangeza committed:
replaced several magic numbers by meaningful constants

hangeza committed:
applied clang-format

hangeza committed:
added toolTips to scan, i2c and map tabs in GUI

hangeza committed:
moved global metatype registrations to main.cpp

hangeza committed:
- changed types of some constants to be more readable
- changed adcSamplesBuffer to STL container

hangeza committed:
changed more constant types to std types

hangeza committed:
Merge branch 'v2.1.0' into dev

hangeza committed:
removed unused includes

hangeza committed:
replaced c array by dynamic array

hangeza committed:
applied clang-format

SimonGLM committed:
Added tooltips to many GUI elements

SimonGLM committed:
Addendum to previous: One tooltip changed

SimonGLM committed:
Addendum to previous: Removed html style tooltips where unnecessary.

SimonGLM committed:
Added 'qwt' to NAMES of QWT_QT5 , fixes build on Arch distros

SimonGLM committed:
Revert "Added 'qwt' to NAMES of QWT_QT5 , fixes build on Arch distros"

This reverts commit f0b742190f42a475eab392476cdcaba691ac5208.

SimonGLM committed:
Added patch file to fix find_library for qwt on Arch

SimonGLM committed:
Fixed typo in tooltip

SimonGLM committed:
Removed ESC to exit.
Use OS methods instead... (I.e. Alt+F4)

marvin5300 committed:
Fixed network discovery for multiple network interfaces and made it more verbose in GUI.

marvin5300 committed:
Switched QHostAddress::Any to QHostAddress::AnyIPv4 in socket bind address for network discovery.
This might fix a bug where network discovery fails due to AnyIPv6 automatically selected.

SimonGLM committed:
Added check to prevent file overwrite

SimonGLM committed:
Fix in NetworkDiscovery qDebug output

SimonGLM committed:
Fixed bug in NetworkDiscovery workflow

SimonGLM committed:
Changed changelogs

SimonGLM committed:
Removed excess qDebug

SimonGLM committed:
Improved info message

marvin5300 committed:
Fixed 0x2a header for NetworkDiscovery data.
Added some validation checks for network datagram in NetworkDiscovery.

marvin5300 committed:
Added check if first word in NetworkDiscovery message is 0x2a.
Added check if data size is at least 4 bytes long.

SimonGLM committed:
Removed unnecessary qDebug of datagram size

SimonGLM committed:
Merge commit '5da24ae77037c3344817bf227996218eee60b371' into v2.1.0
